stablehlo/transforms/StablehloLegalizeQDQToQuantizedOp.cpp

```diff
@@ -0,0 +1,137 @@
+#include "mlir/Dialect/Quant/QuantTypes.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"  // Include for TypeConverter
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "stablehlo/dialect/StablehloOps.h"
+#include "stablehlo/transforms/Passes.h"
+
+namespace mlir {
+namespace stablehlo {
+
+#define GEN_PASS_DEF_STABLEHLOLEGALIZEQDQTOQUANTIZEDOPPASS
+#include "stablehlo/transforms/Passes.h.inc"
+
+namespace {
+
+bool isAnyQuantizedTypes(TypeRange types) {
+  return llvm::any_of(types, [](Type type) {
+    return isa<quant::QuantizedType>(getElementTypeOrSelf(type));
+  });
+}
+
+struct QuantizedStablehloQDQToQuantizedOpConversion
+    : public OpRewritePattern<stablehlo::UniformQuantizeOp> {
+  using OpRewritePattern<stablehlo::UniformQuantizeOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(stablehlo::UniformQuantizeOp quantOp,
+                                PatternRewriter& rewriter) const override {
+    // Matching sequence of ops:
+    //    UniformDequantizeOp -> non-quantized Op -> UniformQuantizeOp.
+    // Start matching from a UniformQuantizeOp (`quantOp`).
+    // Get the Op (`computeOp`) which defines the inputs of the `quantOp`.
+    // Verify all inputs of the `computeOp` are produced by UniformDequantizeOp.
+    // Note: The pass does not delete any prexisting op.
+    auto* computeOp = quantOp->getOperand(0).getDefiningOp();
+    if (!computeOp)
+      return rewriter.notifyMatchFailure(
+          quantOp, "requires operand to be defined by an op");
+
+    if (computeOp->getNumRegions() != 0)
+      return rewriter.notifyMatchFailure(computeOp,
+                                         "ops with regions are not supported");
+
+    if (computeOp->getNumResults() > 1)
+      return rewriter.notifyMatchFailure(
+          computeOp, "ops with variadic results are not supported");
+
+    if (isAnyQuantizedTypes(computeOp->getOperandTypes()))
+      return rewriter.notifyMatchFailure(computeOp,
+                                         "requires non quantized operands");
+
+    // Collect quantized operands and result types to rewrite
+    // All operands and results must be quantized
+    llvm::SmallVector<Value> quantizedComputeOpOperands;
+    for (const Value& operand : computeOp->getOperands()) {
+      auto* definingOp = operand.getDefiningOp();
+      if (!definingOp)
+        return rewriter.notifyMatchFailure(
+            computeOp, "requires operand to be defined by an op");
+
+      auto dequantOp = dyn_cast<stablehlo::UniformDequantizeOp>(definingOp);
+      if (!dequantOp)
+        return rewriter.notifyMatchFailure(
+            definingOp,
+            "requires operand to be defined by an stablehlo.uniform_dequantize "
+            "op");
+
+      quantizedComputeOpOperands.push_back(dequantOp->getOperand(0));
+    }
+    SmallVector<Type> quantizedComputeOpResultTypes(quantOp->getResultTypes());
```

Comment:
nit:Can we just inline this single use variable.?

Reply:
done